### PR TITLE
feat: add a custom user header to `NippyJar`

### PR DESCRIPTION
**Pull request by joshieDo**
* Adds custom user data to `NippyJar`
* Some smol changes from the #4512  review

PR into #4682

### CODE REVIEW

File: crates/storage/nippy-jar/src/compression/zstd.rs [hunk 1]
```diff
@@ -23,6 +23,8 @@ pub enum ZstdState {
 pub struct Zstd {
     /// State. Should be ready before compressing.
     pub(crate) state: ZstdState,
+    /// Compression level. A level of `0` uses zstd's default (currently `3`).
+    pub(crate) level: i32,
```
Comment (mattsse) [inline crates/storage/nippy-jar/src/compression/zstd.rs:27]: can this be unsigned?

File: crates/storage/nippy-jar/src/lib.rs [hunk 2]
```diff
@@ -33,36 +33,68 @@ const NIPPY_JAR_VERSION: usize = 1;
...
+/// Note: that the key (eg. BlockHash) passed to a filter and phf does not need to actually be
+/// stored.
+///
+/// Ultimately, the `freeze` function yields two files: a data file containing both the data and its
+/// configuration, and an index file that houses the offsets and offsets_index.
 #[derive(Debug, Serialize, Deserialize)]
 #[cfg_attr(test, derive(PartialEq))]
-pub struct NippyJar {
-    /// Version
+pub struct NippyJar<H> {
```
Comment (mattsse) [inline crates/storage/nippy-jar/src/lib.rs:65]: this leads to weird API, see `NoJarHeader` helper type

can we use a default value = `()` here?

also, how does this generic affect `NippyJar::load`?

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/storage/nippy-jar/src/lib.rs (hunk 2 above)
Comment (joshieDo) [reply crates/storage/nippy-jar/src/lib.rs]: the user needs to ensure that the type is the same as in `new`

File: crates/storage/nippy-jar/src/compression/zstd.rs (hunk 1 above)
Comment (joshieDo) [reply crates/storage/nippy-jar/src/compression/zstd.rs:27]: the inner lib expects a i32

**Pull request merged**